### DISCUSSION TITLE
Visualize changes in documents, i.e., `git diff`

### DISCUSSION THREAD

**Discussion by lnnrtwttkhn** [quarto-dev/quarto-cli#10351] [Feature Requests]
### Description

Thanks for all your work on Quarto, I really love it.

A feature that would be very cool would be way to **visualize changes in documents**, e.g., leveraging the output of `git diff`.
The main use-case that I regularly have is that I want to be able to show what recently has changed in a document and provide a way to show the revision history to users without them having to know Git or visit the remote repository.
Specifically, some feature that I have in mind:

- [ ] Visualize changes in the same (current) Quarto document
- [ ] Select which change(s) to show: Full history, only latest change, or specific change (i.e., change between two arbitrary commits)
- Visualize changes in *other* Quarto documents (i.e., in `index.qmd` I can visualize the changes in `functions.R`.

Please let me know if I should provide more details. Curious to hear what others think about this or if something along those lines has already been implemented somewhere. Thanks!

**Comment by mcanouil** [comment]
I'm not sure Quarto will do something on that in the future or at least in a near future.

I would suggest you get acquainted with the extension mechanism which is there precisely for you to extend and add such features, see <https://quarto.org/docs/extensions/>.

> if something along those lines has already been implemented somewhere

I am not aware of something like that but you can take a look at all the extensions listed on <https://m.canouil.dev/quarto-extensions/>

**Reply by mcanouil** [reply]
FWIW, how do you translate a markdown git diff to an HTML/LaTeX/Typst diff? What about style diff?
It seems you basically want the Microsoft Word revision mode but for multiple formats. You might see how this is not trivial and cannot really be a priority feature as the amount of work required would be huge.

**Reply by lnnrtwttkhn** [reply]
Thanks @mcanouil for the quick response. Yes, I expected that this would not be a small feature to integrate. This was in the back of my mind for some time and I just wanted to put it out here in case someone else had thought about this before as well. Will look into how to create an extension. Thanks again!

**Reply by mcanouil** [reply]
Thanks for sharing the idea! That's how extensions get created ;)

**Reply by gordonwoodhull** [reply]
It would be easy enough to do a text diff on the qmd and other source files. It's the diff on anything structural or stylistic that would be really difficult.

I'd suggest starting with plain text diff, then add formatting the lines based on whether they start with `+` or `-`.

I've done this in the past (in straight HTML) and it wasn't hard to get decent formatting. Since Quarto is multi-format you'd want to depend on something like https://m.canouil.dev/quarto-highlight-text/ to format the lines, and then use a monospace font. (It's not possible to format code directly right now.)

**Reply by mcanouil** [reply]
You could also leverage the `` ```diff `` code block where lines with `-` are red and with `+` are green.

And beyond the highlight part, a diff needs a reference. How do you define it? The last commit (head)? Then, it means you need to do massive commits if you still want the diff?

Making all this a general framework/workflow is really not trivial and has a lot of design choices.

**Reply by gordonwoodhull** [reply]
Much easier! Here is a trivial Python example, works across formats.

````
```{python}
import subprocess
from IPython.core.display import display, Markdown

diff = subprocess.Popen(['diff', 'a.txt', 'b.txt'], stdout= subprocess.PIPE)
display(Markdown(
  '```diff\n' +
  diff.stdout.read().decode('utf-8') +
  '\n```\n'))
```
````

I agree making a general extension might be complex, seems like the display part is already there!

**Reply by mcanouil** [reply]
I mean displaying the diff of the plain text input inside the document is "easy" but having the corresponding diff such as having the new sentence from the markdown diff be in green in the HTML/Typst/Docx/etc. is way harder and I believe that's the target.
Something close to the track change feature of Microsoft word.

The diff could be added as something close to the "code-tools".

But having the "whole history" means to store the whole git history of a file.
It's really like reimplementing GitHub/GitLab inside Quarto at some level.😅 (I would suggest to teach user to navigate the history there rather than reimplementing the feature)

Edit: adding a link to file history might even be the easiest and most effective way to achieve the desired goal: `https://github.com/<user>/<repo>/commits/main/path/to/file.qmd`

**Reply by bjornasm** [reply]
> (I would suggest to teach user to navigate the history there rather than reimplementing the feature)

Regarding this, sadly - in scientific publishing some journals require tracking changes in the review process.

**Reply by rainer-rq-koelle** [reply]
Such a prose-based feature would be like crossing the rubicon for making qmd output fully workable in day-to-day business work. 
For code chunks, I think we are speaking about another audience ... and this audience is able to handle diffs.

While I was able to pull over many people in my organisation to use (R/)quarto, the day-to-day test is still review of documents and being able to show changes, i.e., emulating the well-knows Word track change. I conquer with @lnnrtwttkhn that getting everyone into checking git diffs, it not sustainable/feasible. 
While a generic feature, being able to show a track record of all changes would be fantastic, I understand the complexity.

From a day-to-day perspective, it would help already, if one could highlight (think the well-known track change side bar in Word) the paras that have changed. I see this more as a summary snapshot based on a "version" of the document, i.e., not necessarily every commit/push diff.
Being able to highlight/change the colour of the changed prose between the version would be the icing on the cake.

I am cognisant that I am using the term "version" here to describe the document after a certain commit (and less all diffs leading up to that point) and then the "delta" to a next "version" (that will typically comprise multiple of commits/diffs).

Obviously, purist will then ask for being able to "remove" such highlights for smaller changes (e.g., adding a comma, correcting a typo), but I can picture this to be a workflow / process step. Fix tiny things first - declare a "version" - include (major) changes to be highlighted. Alternatively, one can work out something like only change colour of version diff if not more than 1 or 2 characters are affected, if more than also highlight bar at para.

Thus, a +1 for such a feature.

**Reply by mcanouil** [reply]
Keep an eye on Quarto 2:

- https://github.com/quarto-dev/quarto-cli/discussions/14323

**Reply by rainer-rq-koelle** [reply]
Jumping with a big smile on my face. You guys are awesome. Thanks for considering some of the day-to-day pain points. Highly appreciated. I am already excited for the rewrite to arrive!

**Reply by cscheid** [reply]
Here's a teaser. You'll notice right away that the styling is really not ideal (in particular the title block), but we're about to merge an attribution feature to Quarto 2, which will work with both `git blame` attribution and automerge-based attribution (from our upcoming collaborative editor)

<img width="1808" height="1380" alt="image" src="https://github.com/user-attachments/assets/657881a7-98c4-4360-9447-a62efa2b58d8" />

**Reply by cscheid** [reply]
(This is from @shikokuchuo's excellent work in the https://github.com/quarto-dev/q2/pull/190 PR)
